### ISSUE TITLE
Can not install cffi using poetry

### ISSUE THREAD

**Issue by Stiegnate** [colesbury/nogil#19]
I just tried to use nogil on my side project, and found out there is an error when installing [cffi](https://foss.heptapod.net/pypy/cffi). Seems like there is a confict between [nogil](https://github.com/colesbury/nogil/blob/nogil/Include/object.h#L119) and [CPython](https://github.com/python/cpython/blob/main/Include/object.h#L105).

Origin error message as following,

```
Installing cffi (1.15.0): Failed

  EnvCommandError

  Command ['/workspace/.venv/bin/pip', 'install', '--no-deps', 'file:///root/.cache/pypoetry/artifacts/20/b6/2f/d86d84b144949a7c7435d77e30e4731b66f07f492faac0f75b18bebca7/cffi-1.15.0.tar.gz'] errored with the following return code 1, and output: 
  Processing /root/.cache/pypoetry/artifacts/20/b6/2f/d86d84b144949a7c7435d77e30e4731b66f07f492faac0f75b18bebca7/cffi-1.15.0.tar.gz
  Building wheels for collected packages: cffi
    Building wheel for cffi (setup.py): started
    Building wheel for cffi (setup.py): finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: /workspace/.venv/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-ned2bb9p/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-ned2bb9p/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-fies3lsc
         cwd: /tmp/pip-req-build-ned2bb9p/
    Complete output (36 lines):
    running bdist_wheel
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.9
    creating build/lib.linux-x86_64-3.9/cffi
    copying cffi/commontypes.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/cparser.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/model.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/cffi_opcode.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/recompiler.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/pkgconfig.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/api.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/verifier.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/backend_ctypes.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/vengine_cpy.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/__init__.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/setuptools_ext.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/ffiplatform.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/vengine_gen.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/lock.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/error.py -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/_cffi_include.h -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/parse_c_type.h -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/_embedding.h -> build/lib.linux-x86_64-3.9/cffi
    copying cffi/_cffi_errors.h -> build/lib.linux-x86_64-3.9/cffi
    running build_ext
    building '_cffi_backend' extension
    creating build/temp.linux-x86_64-3.9
    creating build/temp.linux-x86_64-3.9/c
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/workspace/.venv/include -I/usr/local/include/python3.9 -c c/_cffi_backend.c -o build/temp.linux-x86_64-3.9/c/_cffi_backend.o
    c/_cffi_backend.c: In function ‘get_unique_type’:
    c/_cffi_backend.c:4638:20: error: ‘PyObject’ {aka ‘struct _object’} has no member named ‘ob_refcnt’
     4638 |     ((PyObject *)x)->ob_refcnt--;
          |                    ^~
    error: command 'gcc' failed with exit status 1
    ----------------------------------------
    ERROR: Failed building wheel for cffi
    Running setup.py clean for cffi
  Failed to build cffi
  Installing collected packages: cffi
      Running setup.py install for cffi: started
      Running setup.py install for cffi: finished with status 'error'
      ERROR: Command errored out with exit status 1:
       command: /workspace/.venv/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-ned2bb9p/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-ned2bb9p/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-e26e6dwz/install-record.txt --single-version-externally-managed --compile --install-headers /workspace/.venv/include/site/python3.9/cffi
           cwd: /tmp/pip-req-build-ned2bb9p/
      Complete output (36 lines):
      running install
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.9
      creating build/lib.linux-x86_64-3.9/cffi
      copying cffi/commontypes.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/cparser.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/model.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/cffi_opcode.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/recompiler.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/pkgconfig.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/api.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/verifier.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/backend_ctypes.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/vengine_cpy.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/__init__.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/setuptools_ext.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/ffiplatform.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/vengine_gen.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/lock.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/error.py -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/_cffi_include.h -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/parse_c_type.h -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/_embedding.h -> build/lib.linux-x86_64-3.9/cffi
      copying cffi/_cffi_errors.h -> build/lib.linux-x86_64-3.9/cffi
      running build_ext
      building '_cffi_backend' extension
      creating build/temp.linux-x86_64-3.9
      creating build/temp.linux-x86_64-3.9/c
      gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/workspace/.venv/include -I/usr/local/include/python3.9 -c c/_cffi_backend.c -o build/temp.linux-x86_64-3.9/c/_cffi_backend.o
      c/_cffi_backend.c: In function ‘get_unique_type’:
      c/_cffi_backend.c:4638:20: error: ‘PyObject’ {aka ‘struct _object’} has no member named ‘ob_refcnt’
       4638 |     ((PyObject *)x)->ob_refcnt--;
            |                    ^~
      error: command 'gcc' failed with exit status 1
      ----------------------------------------
  ERROR: Command errored out with exit status 1: /workspace/.venv/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-ned2bb9p/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-ned2bb9p/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-e26e6dwz/install-record.txt --single-version-externally-managed --compile --install-headers /workspace/.venv/include/site/python3.9/cffi Check the logs for full command output.
  WARNING: You are using pip version 21.2.4; however, version 21.3 is available.
  You should consider upgrading via the '/workspace/.venv/bin/python -m pip install --upgrade pip' command.
  

  at /usr/local/lib/python3.9/site-packages/poetry/utils/env.py:1183 in _run
      1179│                 output = subprocess.check_output(
      1180│                     cmd, stderr=subprocess.STDOUT, **kwargs
      1181│                 )
      1182│         except CalledProcessError as e:
    → 1183│             raise EnvCommandError(e, input=input_)
      1184│ 
      1185│         return decode(output)
      1186│ 
      1187│     def execute(self, bin, *args, **kwargs):
```

**Comment by ApproximateIdentity** [comment]
Judging by this part of the error message

```
      gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/workspace/.venv/include -I/usr/local/include/python3.9 -c c/_cffi_backend.c -o build/temp.linux-x86_64-3.9/c/_cffi_backend.o
      c/_cffi_backend.c: In function ‘get_unique_type’:
      c/_cffi_backend.c:4638:20: error: ‘PyObject’ {aka ‘struct _object’} has no member named ‘ob_refcnt’
       4638 |     ((PyObject *)x)->ob_refcnt--;
            |                    ^~
      error: command 'gcc' failed with exit status 1
```

it looks like the issue is that cffi is accessing the `ob_refcnt` field that was removed in this commit:

https://github.com/colesbury/nogil/commit/24d53bbd501665b6f86cf9a95bf8b3e70ef0c62b

If you look at the "Compatibility" section of his doc here https://docs.google.com/document/d/18CXhDb1ygxg-YXNBJNzfzZsDFosB5e6BfnXLlejd9l0/edit#heading=h.fukli2k90k4 you see that he explicitly mentions this issue:

```
To compile successfully, some C API extensions will need minor modifications, typically
replacement of direct access of PyObject reference count fields with the Py_REFCNT macro.
```

So I guess the first step would be to grab the cffi source code and see if that change fixes things.

**Comment by ApproximateIdentity** [comment]
Yeah I just checked that if you change this line https://foss.heptapod.net/pypy/cffi/-/blob/branch/release-1.15/c/_cffi_backend.c#L4638 from `((PyObject *)x)->ob_refcnt--;` to `Py_DECREF(x);`, then it will compile and install successfully.

That said, if you read the comment above that line here https://foss.heptapod.net/pypy/cffi/-/blob/branch/release-1.15/c/_cffi_backend.c#L4635-4637 you see that they are purposely not using the macro there. Judging by the comment, it looks like the reason isn't important unless you're doing a debug build, but it's good to be aware. They aren't using the field directly by accident; it is done on purpose.

**Comment by colesbury** [comment]
There is a compatible version of `cffi` (with [a change](https://github.com/colesbury/cffi/commit/8f2f756a46f12180fc812903502d4d71c23f31f4) similar to the one described by @ApproximateIdentity) that should be pulled by `pip`, but doesn't seem to be in your case. How are you installing `cffi`?

If you are using "pip install", can you post the output of `pip3 install -vvv cffi`?

For example:

```
pip3 install -vvv cffi 2>&1 | tee pip.log
```

(And then share the contents of pip.log)

**Comment by ApproximateIdentity** [comment]
> There is a compatible version of `cffi` (with [a change](https://github.com/colesbury/cffi/commit/8f2f756a46f12180fc812903502d4d71c23f31f4) similar to the one described by @ApproximateIdentity) that should be pulled by `pip`, but doesn't seem to be in your case. How are you installing `cffi`?

That link gives me a 404 when I click it and I don't see `cffi` listed as a repo under your name. Is it possibly marked set as private?

**Comment by colesbury** [comment]
@ApproximateIdentity, sorry it was marked as private. I've made it public now.

**Comment by Stiegnate** [comment]
> There is a compatible version of `cffi` (with [a change](https://github.com/colesbury/cffi/commit/8f2f756a46f12180fc812903502d4d71c23f31f4) similar to the one described by @ApproximateIdentity) that should be pulled by `pip`, but doesn't seem to be in your case. How are you installing `cffi`?

I use [poetry](https://github.com/python-poetry/poetry) as Python package management tool, and `cffi==1.15.0` is one of nested dependency.

> If you are using "pip install", can you post the output of `pip3 install -vvv cffi`?
> 

Here is `pip.log`. It works in default python.

```
Using pip 21.2.4 from /usr/local/lib/python3.9/site-packages/pip (python 3.9)
Non-user install because site-packages writeable
Created temporary directory: /tmp/pip-ephem-wheel-cache-obh8v0v4
Created temporary directory: /tmp/pip-req-tracker-1dzi9nga
Initialized build tracking at /tmp/pip-req-tracker-1dzi9nga
Created build tracker: /tmp/pip-req-tracker-1dzi9nga
Entered build tracker: /tmp/pip-req-tracker-1dzi9nga
Created temporary directory: /tmp/pip-install-lvg45xxu
Looking in indexes: https://d1yxz45j0ypngg.cloudfront.net/, https://pypi.org/simple
Requirement already satisfied: cffi in /usr/local/lib/python3.9/site-packages (1.14.6)
Requirement already satisfied: pycparser in /usr/local/lib/python3.9/site-packages (from cffi) (2.20)
Created temporary directory: /tmp/pip-unpack-387g7ipu
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Removed build tracker: '/tmp/pip-req-tracker-1dzi9nga'
```

**Comment by colesbury** [comment]
If you use `poetry`, you'll have to add a custom repository and make sure to use packages available in that repository:

```
[[tool.poetry.source]]
name = "nogil"
url = "https://d1yxz45j0ypngg.cloudfront.net/"
```

```
[tool.poetry.dependencies]
cffi = "1.14.6"
```

(Only cffi 1.14.6 is available for nogil Python; not 1.15.0)

(and then run `poetry update`)
